PNG converted to HEIC on file upload seemingly not accepted by image/* wildcard

karlcow opened the pull request:
#### 4a46daaf002f2033d4e7425f4690c8fc759a606a
<pre>
PNG converted to HEIC on file upload seemingly not accepted by image/* wildcard
<a href="https://bugs.webkit.org/show_bug.cgi?id=303803">https://bugs.webkit.org/show_bug.cgi?id=303803</a>
<a href="https://rdar.apple.com/166124206">rdar://166124206</a>

Reviewed by Darin Adler.

When image/*, image/heic is sent in a form with an input

&lt;input type=&quot;file&quot; multiple
  accept=&quot;image/*, image/heic&quot;
  id=&quot;file&quot;&gt;

and the file being uploaded is a PNG image. We do not want to convert it
to HEIC. We want to keep it as PNG.

* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::findImagesForTranscoding):

Canonical link: <a href="https://commits.webkit.org/305283@main">https://commits.webkit.org/305283@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/375107ec2d609ac8d0eada026baff00403c29e69

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137997 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10362 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49382 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146068 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/90972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/580eed48-d883-44cb-be98-7a6819de9106) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/139870 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11063 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10509 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105534 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/90972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/298b880a-3b48-495e-a95a-4efd039a5bda) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140942 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8258 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123724 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86384 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7872 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5622 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6349 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117269 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41900 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148776 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10044 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42459 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113935 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10061 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8479 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114267 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7809 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119987 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/64768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21239 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10090 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37964 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9821 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73658 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/10031 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9882 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -99,10 +99,17 @@ static String transcodeImage(const String& path, const String& destinationUTI, c
 
 Vector<String> findImagesForTranscoding(const Vector<String>& paths, const Vector<String>& allowedMIMETypes)
 {
+    // FIXME: Do we really want to allow all image formats or just a list of them.
```
Can you clarify what this FIXME means? Are you suggesting that there are cases where we should still transcode even when `image/*` is specified?

karlcow replied:
@pxlcoder 

I was wondering about two cases:
* `image/svg`, but maybe it doesn't matter. 
* `image/foobar` (a not supported content type), 

What should Safari do? Maybe it doesn't matter. 
Happy to hear any opinions about that.

pxlcoder replied:
Please correct me if I'm misunderstanding your cases, but:
- `accept=image/svg`: ideally we should only accept SVGs (and make other files unselectable), since we cannot transcode to SVG
- `accept=image/foobar`: all files should be unselectable

darinadler replied:
I think that @karlcow is asking a question like this: _If the source image is an SVG and the accept attribute lists `image/*`, should WebKit convert it to a PNG or upload an SVG file?_

darinadler replied:
And, similarly: _If the source image is some kind of exotic image type that didn’t exist when the website was designed and the HTML specification was written, but it happens to have an assigned MIME type of the form `image/something`, then should WebKit convert it to a PNG or upload the file?_

pxlcoder replied:
I think we should not transcode in both those cases (existing behavior).
- I believe that `/*` supporting forward compatibility is a feature
- For SVG specifically, I don't think it makes sense to transcode as a long supported format

One reason to reconsider might be if there are known compat issues.

karlcow replied:
Thanks @darinadler for making my thoughts clearer. :)

karlcow replied:
OK I tested a couple of scenarios to make sure about what was happening and if the patch was working as expected. 


| Input | Image | Safari | MiniBrowser Patched | Firefox | Chrome |
|-|-|-|-|-|-|
| `<input type="file">`| PNG | `image/png` | same | `image/png` | `image/png` |
| `<input type="file">`| SVG | `image/svg+xml` | same | `image/svg+xml` | `image/svg+xml` |
| `<input type="file">`| HEIC | `image/heic` | same | `image/heic` | `image/heic` |
| `<input type="file">`| text | `text/plain` | same | `text/plain` | `text/plain` |
| `<input type="file">`| unknown | `application/octet-stream` | same | `application/octet-stream` | `application/octet-stream` |
| `<input type="file" accept="image/*">`| PNG | `image/png` | same | `image/png` | `image/png` |
| `<input type="file" accept="image/*">`| SVG | `image/svg+xml` | same | `image/svg+xml` | `image/svg+xml` |
| `<input type="file" accept="image/*">`| HEIC | `image/heic` | same | `image/heic` | `image/heic` |
| `<input type="file" accept="image/*">`| text | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/*">`| unknown | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/png">`| PNG | `image/png` | same | `image/png` | `image/png` |
| `<input type="file" accept="image/png">`| SVG | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/png">`| HEIC | **`image/png`** | same | not selectable | not selectable |
| `<input type="file" accept="image/png">`| text | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/png">`| unknown | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/svg">`| PNG | `image/png` | same | `image/png` | `image/png` |
| `<input type="file" accept="image/svg">`| SVG | `image/svg+xml` | same | `image/svg+xml` | `image/svg+xml` |
| `<input type="file" accept="image/svg">`| HEIC | `image/heic` | same | `image/heic` | `image/heic` |
| `<input type="file" accept="image/svg">`| text | `text/plain` | same | `text/plain` | `text/plain` |
| `<input type="file" accept="image/svg">`| unknown | `application/octet-stream` | same | `application/octet-stream` | `application/octet-stream` |
| `<input type="file" accept="image/svg+xml">`| PNG | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/svg+xml">`| SVG | `image/svg+xml` | same | `image/svg+xml` | `image/svg+xml` |
| `<input type="file" accept="image/svg+xml">`| HEIC | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/svg+xml">`| text | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/svg+xml">`| unknown | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/*, image/png">`| PNG | `image/png` | same | `image/png` | `image/png` |
| `<input type="file" accept="image/*, image/png">`| SVG | `image/svg+xml` | same | `image/svg+xml` | `image/svg+xml` |
| `<input type="file" accept="image/*, image/png">`| HEIC | **`image/png`** | **`image/heic`** | `image/heic` | `image/heic` |
| `<input type="file" accept="image/*, image/png">`| text | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/*, image/png">`| unknown | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/png, image/*">`| PNG | `image/png` | same | `image/png` | `image/png` |
| `<input type="file" accept="image/png, image/*">`| SVG | `image/svg+xml` | same | `image/svg+xml` | `image/svg+xml` |
| `<input type="file" accept="image/png, image/*">`| HEIC | **`image/png`** | **`image/heic`** | `image/heic` | `image/heic` |
| `<input type="file" accept="image/png, image/*">`| text | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/png, image/*">`| unknown | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg, image/png">`| PNG | `image/png` | same | `image/png` | `image/png` |
| `<input type="file" accept="image/jpeg, image/png">`| SVG | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg, image/png">`| HEIC | **`image/jpeg`** | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg, image/png">`| text | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg, image/png">`| unknown | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg">`| PNG | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg">`| SVG | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg">`| HEIC | **`image/jpeg`** | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg">`| text | not selectable | same | not selectable | not selectable |
| `<input type="file" accept="image/jpeg">`| unknown | not selectable | same | not selectable | not selectable |

karlcow replied:
it fixes the issue.

pxlcoder reviewed (commented):
This should be layout-testable, you can look at the following tests for inspiration:
- `LayoutTests/fast/forms/file/entries-api/image-no-transcode-open-panel.html`
- `LayoutTests/fast/forms/file/entries-api/image-transcode-open-panel.html`

darinadler commented on Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
```diff
@@ -99,10 +99,17 @@ static String transcodeImage(const String& path, const String& destinationUTI, c
 
 Vector<String> findImagesForTranscoding(const Vector<String>& paths, const Vector<String>& allowedMIMETypes)
 {
+    // FIXME: Do we really want to allow all image formats or just a list of them.
```
I think the real practical short term question we should be asking is: Are there cases that this makes worse in practice, where the transcoding is helpful to web browser users, and this code change will make us stop doing it? A more conservative change would make an exception only for some specific image types, which would be less elegant, but we could be more sure we didn’t remove transcoding in a case where it’s useful.

darinadler reviewed (approved):
This looks right to me. We could consider a more conservative change if we’re worried about removing some useful transcoding cases by accident.

karlcow committed:
PNG converted to HEIC on file upload seemingly not accepted by image/* wildcard
https://bugs.webkit.org/show_bug.cgi?id=303803
rdar://166124206

Reviewed by Darin Adler.

When image/*, image/heic is sent in a form with an input

<input type="file" multiple
  accept="image/*, image/heic"
  id="file">

and the file being uploaded is a PNG image. We do not want to convert it
to HEIC. We want to keep it as PNG.

* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::findImagesForTranscoding):

Canonical link: https://commits.webkit.org/305283@main
